File: src/offboard_control/offboard_control/vio_bridge.py
#!/usr/bin/env python3
"""
vio_bridge.py - Converts VINS-Fusion odometry (ENU) to PX4 VehicleVisualOdometry (NED)
Required for PX4 to use VIO as external vision.

VINS outputs: /vins_estimator/camera_pose (PoseStamped) in ENU
PX4 expects: /fmu/in/vehicle_visual_odometry (VehicleOdometry) in NED, with covariance

Also publishes TF map->base_link and odom.

For real drone: set EKF2_AID_MASK to use vision.
"""
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Odometry
from px4_msgs.msg import VehicleOdometry
import math
import numpy as np
from scipy.spatial.transform import Rotation as R

class VIOBridge(Node):
    def __init__(self):
        super().__init__('vio_bridge')
        self.declare_parameter('input_topic', '/vins_estimator/camera_pose')
        self.declare_parameter('output_topic', '/fmu/in/vehicle_visual_odometry')

        input_topic = self.get_parameter('input_topic').value
        output_topic = self.get_parameter('output_topic').value

        self.vio_sub = self.create_subscription(PoseStamped, input_topic, self.vio_callback, 10)
        # Alternative odom topic
        self.odom_sub = self.create_subscription(Odometry, '/vins_estimator/odometry', self.odom_callback, 10)

        self.visual_odom_pub = self.create_publisher(VehicleOdometry, output_topic, 10)
        self.pose_pub = self.create_publisher(PoseStamped, '/mavros/vision_pose/pose', 10)

        self.last_pose = None
        self.get_logger().info(f"VIO Bridge: {input_topic} -> {output_topic}")

    def pose_to_ned(self, pose_enu):
        # ENU to NED conversion
        # ENU: x east, y north, z up
        # NED: x north, y east, z down
        x_enu = pose_enu.position.x
        y_enu = pose_enu.position.y
        z_enu = pose_enu.position.z

        x_ned = y_enu
        y_ned = x_enu
        z_ned = -z_enu

        # Quaternion ENU to NED: need to rotate
        # Simplified: ENU to NED is 90 deg yaw + flip
        # For accurate conversion, use rotation matrix
        q_enu = [pose_enu.orientation.x, pose_enu.orientation.y, pose_enu.orientation.z, pose_enu.orientation.w]
        r_enu = R.from_quat(q_enu)
        # ENU to NED rotation matrix
        # Transform: R_ned = R_enu_to_ned * R_enu
        # R_enu_to_ned = [[0,1,0],[1,0,0],[0,0,-1]]
        mat_enu_to_ned = np.array([[0,1,0],[1,0,0],[0,0,-1]])
        r_ned = R.from_matrix(mat_enu_to_ned @ r_enu.as_matrix())
        q_ned = r_ned.as_quat()  # x,y,z,w

        return (x_ned, y_ned, z_ned), q_ned

    def vio_callback(self, msg):
        # Convert to VehicleOdometry
        (x_ned, y_ned, z_ned), q_ned = self.pose_to_ned(msg.pose)

        odom_msg = VehicleOdometry()
        odom_msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        odom_msg.timestamp_sample = odom_msg.timestamp
        odom_msg.pose_frame = VehicleOdometry.POSE_FRAME_NED
        odom_msg.position = [x_ned, y_ned, z_ned]
        odom_msg.q = [float(q_ned[3]), float(q_ned[0]), float(q_ned[1]), float(q_ned[2])]  # w,x,y,z for PX4
        odom_msg.velocity = [0.0, 0.0, 0.0]
        odom_msg.angular_velocity = [0.0, 0.0, 0.0]
        odom_msg.position_variance = [0.1, 0.1, 0.1]
        odom_msg.orientation_variance = [0.1, 0.1, 0.1]
        odom_msg.velocity_variance = [0.1, 0.1, 0.1]

        self.visual_odom_pub.publish(odom_msg)

        # Also publish as PoseStamped for mapping
        self.pose_pub.publish(msg)
        self.last_pose = msg

    def odom_callback(self, msg):
        # Similar conversion for Odometry
        pose_stamped = PoseStamped()
        pose_stamped.header = msg.header
        pose_stamped.pose = msg.pose.pose
        self.vio_callback(pose_stamped)

def main(args=None):
    rclpy.init(args=args)
    node = VIOBridge()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/offboard_control_cpp/src/vio_bridge.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <px4_msgs/msg/vehicle_odometry.hpp>
#include <Eigen/Dense>
#include <cmath>

/*
 * vio_bridge.cpp - C++ fast ENU->NED conversion for PX4 VIO
 * Uses Eigen for quaternion math, ~5x faster than Python scipy
 */

class VIOBridge : public rclcpp::Node {
public:
  VIOBridge() : Node("vio_bridge") {
    this->declare_parameter("input_topic", "/vins_estimator/camera_pose");
    this->declare_parameter("output_topic", "/fmu/in/vehicle_visual_odometry");

    std::string input = this->get_parameter("input_topic").as_string();
    std::string output = this->get_parameter("output_topic").as_string();

    vio_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(input, 10, std::bind(&VIOBridge::vioCb, this, std::placeholders::_1));
    odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>("/vins_estimator/odometry", 10, std::bind(&VIOBridge::odomCb, this, std::placeholders::_1));

    odom_pub_ = this->create_publisher<px4_msgs::msg::VehicleOdometry>(output, 10);
    pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/mavros/vision_pose/pose", 10);

    RCLCPP_INFO(this->get_logger(), "C++ VIO Bridge %s -> %s", input.c_str(), output.c_str());
  }

private:
  void vioCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
    // ENU to NED: x_ned=y_enu, y_ned=x_enu, z_ned=-z_enu
    double x_enu=msg->pose.position.x, y_enu=msg->pose.position.y, z_enu=msg->pose.position.z;
    double x_ned=y_enu, y_ned=x_enu, z_ned=-z_enu;

    // Quaternion conversion using rotation matrix
    Eigen::Quaterniond q_enu(msg->pose.orientation.w, msg->pose.orientation.x, msg->pose.orientation.y, msg->pose.orientation.z);
    Eigen::Matrix3d R_enu = q_enu.toRotationMatrix();
    Eigen::Matrix3d R_enu_to_ned;
    R_enu_to_ned << 0,1,0, 1,0,0, 0,0,-1;
    Eigen::Matrix3d R_ned = R_enu_to_ned * R_enu;
    Eigen::Quaterniond q_ned(R_ned);
    q_ned.normalize();

    px4_msgs::msg::VehicleOdometry odom;
    odom.timestamp = this->now().nanoseconds()/1000;
    odom.timestamp_sample = odom.timestamp;
    odom.pose_frame = px4_msgs::msg::VehicleOdometry::POSE_FRAME_NED;
    odom.position = {static_cast<float>(x_ned), static_cast<float>(y_ned), static_cast<float>(z_ned)};
    odom.q = {static_cast<float>(q_ned.w()), static_cast<float>(q_ned.x()), static_cast<float>(q_ned.y()), static_cast<float>(q_ned.z())};
    odom.velocity = {0,0,0};
    odom.angular_velocity = {0,0,0};
    odom.position_variance = {0.1,0.1,0.1};
    odom.orientation_variance = {0.1,0.1,0.1};
    odom.velocity_variance = {0.1,0.1,0.1};

    odom_pub_->publish(odom);
    pose_pub_->publish(*msg);
  }

  void odomCb(const nav_msgs::msg::Odometry::SharedPtr msg) {
    geometry_msgs::msg::PoseStamped ps;
    ps.header=msg->header;
    ps.pose=msg->pose.pose;
    vioCb(std::make_shared<geometry_msgs::msg::PoseStamped>(ps));
  }

  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr vio_sub_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
  rclcpp::Publisher<px4_msgs::msg::VehicleOdometry>::SharedPtr odom_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<VIOBridge>());
  rclcpp::shutdown();
  return 0;
}
